### PR TITLE
Blogpost: The evolution of SciPy CLI

### PR DESCRIPTION
**Pull request by sayantikabanik**
## Text styling

- [X] The blog is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [X] If there are headers, they use the proper header tags in order to do so (with only one level-one header).
- [X] All links describe where they link to (for example, check the [Quansight labs website](https://labs.quansight.org/)).
- [X] Any kind of styling that the author uses (for example, bold for emphasis) is consistent throughout the blog.

## Non-text contents

- [X] All content is represented as text (for example, images need alt text and videos need captions or descriptive transcripts).
- [X] If there are emojis, there are not more than three in a row.
- [X] Don't use [flashing gifs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [X] If it were to be read as plain text, the blog still makes sense and no information is missing.

### CODE REVIEW

File: posts/2022/05/the-evolution-of-scipy-cli.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        },
+    ],
+}
+```
+
+#### 🎥 Devloper CLI in action
+
+<script id="asciicast-U9l9VvklvEjXdEi1xYS4A7u5M" src="https://asciinema.org/a/U9l9VvklvEjXdEi1xYS4A7u5M.js" async></script>
+
+#### Current list of implemented tasks
```
Comment (rgommers) [inline posts/2022/05/the-evolution-of-scipy-cli.md:113]: It needs some introductory sentence before the bullet list here.

File: posts/2022/05/the-evolution-of-scipy-cli.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+.. type: text
+.. previewimage:
+-->
+
+#### 🤔 What is a command-line interface (CLI)?
+
+Imagine a situation where there is a massive system with various tools and functionality. Every functionality requires a special command or an input from the user, CLI is designed to tackle such situations. Like a catalog it lists all the options present, helping the user to navigate. A CLI can be a simple list or a complex GUI based application, for example, shown below.
+
+<p align="center">
+  <img src="https://user-images.githubusercontent.com/17350312/166225781-3f2cfd4d-49a5-4e30-a3bf-841552d7338c.jpg" />
```
[the remark this answers is not in the document]
Comment (rgommers) [reply posts/2022/05/the-evolution-of-scipy-cli.md:19]: How about just using the screenshot of the new CLI that we also used on Twitter? That's way nicer.

File: posts/2022/05/the-evolution-of-scipy-cli.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+- Listing the args/options for a task: `python do.py <task_name> --help`
+
+**Outcomes**
+
+1. 📜 Self documentation and hierarchical help option
+2. 🧭 Easy to navigate and intuitive interface 
+3. ⏩ Clear and concise examples to get started quickly
+4. ⏱ Reduction in the time spent navigating documentation
+
+With a great start comes possibilities. In the coming weeks, the CLI will become more mature and stable. After we receive wider usage and acceptance from the community, support for `dev.py` and `runtests.py` will be paused and `do.py` will be renamed to `dev.py`. The user documentation for the CLI components and usage will be made available for clear and concise understanding.
```
Comment (rgommers) [inline posts/2022/05/the-evolution-of-scipy-cli.md:154]: Other next steps: `pydevtool` work that Eduardo is doing to make things more reusable. Adding support for other features, like running GitHub CI jobs via Act.

**Review by rgommers — COMMENTED** [review]
Looks like a good start @sayantikabanik. Here is an initial review.

File: posts/2022/05/the-evolution-of-scipy-cli.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
+<!--
+.. title: The evolution of SciPy CLI
```
[the remark this answers is not in the document]
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-scipy-cli.md:2]: The word `evolution` communicates the journey from `runtests.py`- `dev.py`-`do.py` (In my viewpoint), was hoping to have that in the blog post title.

File: posts/2022/05/the-evolution-of-scipy-cli.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-scipy-cli.md:19]: Was trying to find an image communicating the `CLI` vs `CLI+GUI` (the visually nicer version). I shall update this one

File: posts/2022/05/the-evolution-of-scipy-cli.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+.. category:
+.. link:
+.. description:
+.. type: text
+.. previewimage:
+-->
+
+#### 🤔 What is a command-line interface (CLI)?
+
+Imagine a situation where there is a massive system with various tools and functionality. Every functionality requires a special command or an input from the user, CLI is designed to tackle such situations. Like a catalog it lists all the options present, helping the user to navigate. A CLI can be a simple list or a complex GUI based application, for example, shown below.
```
[the remark this answers is not in the document]
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-scipy-cli.md:16]: I shall rephrase the wordings, thanks for pointing this out.

File: posts/2022/05/the-evolution-of-scipy-cli.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+</p>
+
+Now that we understand what a `CLI` is, how about we dive into the world of `SciPy`?
+
+#### 🤓 SciPy CLI journey
+
+An open-source project undergoes multiple iterations, and each iteration adds layers of tools and functionalities to incorporate. Right from the initial build, test, benchmarking, release notes, and beyond. Contribution guidelines start spanning across multiple pages, exponentially increasing efforts to maintain. 
+
+The two homegrown CLI for SciPy are `runtests.py` and `dev.py` (for meson) built using python tooling `argparse`. 
+Both of these toolings have multiple conditional blocks to achieve the desired outcome. Not only it is difficult to maintain, overall readability decreases drastically. The documentation runs into an infinite loop of updates, requiring additional efforts from maintainers. Another issue that remains is the lack of command bundling. A combined unit of commands linking to a single objective helps reduce confusion, very helpful for new contributors (like me :D).
```
[the remark this answers is not in the document]
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-scipy-cli.md:29]: I meant to communicate `command grouping`. like the `release task` group  [`release_notes`, `release_authors`].  Have rephrased the wordings.

File: posts/2022/05/the-evolution-of-scipy-cli.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+#### Planning and objective
+Like any development activity, the goal was to experiment with tools available `doit` and `typer` were the first ones we picked. The 02 components of our interests were a `task runner`, and a ` command-line interface tool`.  
+
+As a starting point, I began experimenting with existing `dev.py` options, wrapped around multiple composite `doit` and `typer` tasks. Both the [doit POC](https://github.com/sayantikabanik/scipy/blob/cli_poc/dodo.py) and [typer POC](https://github.com/sayantikabanik/scipy/blob/cli_poc/cli.py) were built with similar principles. 
+
+As I progressed with the development of POCs using both tools, I experienced certain shortcomings. A Better way to integrate an external library for exposing a CLI was the missing piece. **Eduardo** author of doit developed an architecture combining the core elements of `doit` along with `click`. The journey in detail is captured under [issue-#133](https://github.com/rgommers/scipy/issues/133).
+
+#### 💁🏽‍♀️  More about the architecture and core components
+
```
[the remark this answers is not in the document]
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-scipy-cli.md:41]: I have included the explanation (one-liner). 
I think doit's advanced functionalities are not yet fully utilised, also the task/ file dependency usage is limited to a few tasks.

File: posts/2022/05/the-evolution-of-scipy-cli.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply posts/2022/05/the-evolution-of-scipy-cli.md:2]: yes indeed - sorry, I wasn't recommending to delete "evolution", just that it was missing "the"

**Comment by noatamir** [comment]
Happy to! by when -ish? 😇

**Comment by rgommers** [comment]
Thanks @noatamir. Not super urgent, but within the next week would be great - and then squash-merge if it all looks good. Or if you prefer I do the final check & merge, then by Friday (I'm on holiday from Sat onwards).

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+.. category:
+.. link:
+.. description:
+.. type: text
+.. previewimage:
+-->
+
+#### 🤔 What is a command-line interface (CLI)?
+
+Imagine a situation where there is a massive system with various tools and functionality. Every functionality requires a special command or an input from the user, CLI is designed to tackle such situations. Like a catalog it lists all the options present, helping the user to navigate.
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:16]:
```suggestion
Imagine a situation, where there is a massive system with various tools and functionalities, and where every functionality requires a special command or an input from the user. A CLI is designed to tackle such situations. Like a catalog or menu, it lists all the options available, thus helping the user to navigate a complex system.
```

Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:16]: tweaked it a bit for readability mostly. But it's just style, I'm not sure if we have guidelines for comma usage?! So if you don't like it better ignore it. 
I think functionalities should be plural though.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+  <img alt="CLI example" src="https://user-images.githubusercontent.com/17350312/166633508-a2795c44-30bc-4a5b-8043-65beab71d31f.png" />
+</p>
+
+Now that we understand what a `CLI` is, how about we dive into the world of `SciPy`?
+
+<!-- TEASER_END -->
+
+#### 🤓 SciPy CLI journey
+
+An open-source project undergoes multiple iterations, and each iteration adds layers of tools and functionalities to incorporate. Right from the initial build, test, benchmarking, release notes, and beyond. Contribution guidelines start spanning across multiple pages, exponentially increasing efforts to maintain.
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:28]: I think the word incorporate in  the first sentence doesn't quite make sense to me. Maybe `to it` simply?

The next line is also  a bit tricky. I think it's missing a verb? or maybe it's supposed to connec to the next one somehow? but on it's own I can't quite make sense of it...

For the last line I'd make suggest a tweak: `Over time, the contribution guides start to span over more and more pages, as the over all effort to maintain the project grows exponentially" 
(I don't actually know if the growth is exponential, I'm taking your lead here 😉 )

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+The two homegrown CLI for SciPy are `runtests.py`(for distutils-based builds) and `dev.py`(for Meson-based builds) built using python tooling `argparse`.
+Both of these toolings have multiple conditional blocks to achieve the desired outcome. Not only it is difficult to maintain, overall readability decreases drastically. The documentation runs into an infinite loop of updates, requiring additional efforts from maintainers. Another issue that remains is the lack of task grouping. A group of tasks linking to a single objective helps reduce confusion, super helpful for new contributors (like me :D).
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:31]: Is `toolings` SciPy lingo or some term I'm not aware of? if not I word amend it to `tools`. 
the context for mentioning the conditionals isn't super clear to me. Why is this piece of  information interesting/relevant in an already information heavy paragraph?  If you think it's interesting for people who have knowledge I lack, do keep it. If you think it's interesting for new contributors , maybe add some insight as to why, like you do so well later on. If it's not critical, suggest to edit to `Both of these tools are hard to maintain, their overall readability decreases  ...`

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+Hence the idea of a developer command-line interface (CLI) was born. Easing the development experience with an intuitive and informative CLI. Removing dependency on legacy tooling like `paver` also added great value to the overall experience. Additional details could be found under [issue-#15489](https://github.com/scipy/scipy/issues/15489)
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:33]:
```suggestion
Thus the idea of a developer command-line interface (CLI) was born, easing the development experience with an intuitive and informative CLI. In addition, we would be removing dependency on legacy tooling like `paver`, which would also add great value to the overall experience. Additional details could be found under [issue-#15489](https://github.com/scipy/scipy/issues/15489)
```

Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:33]: Thus sounds better imo. this is mostly readability again. content looks great.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+
+#### ✍️ Planning and objective
+Like any development activity, the goal was to experiment with tools available [doit](https://pydoit.org/) and [Typer](https://typer.tiangolo.com/) were the first ones we picked. The two components of our interests were a *task runner*, and a *command-line interface tool*. doit satisfied the requirements of a task runner along with added functionality like maintaining a task dependency graph as a `DAG`. While `Typer` is quick to get started with building CLI applications.
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:36]:
```suggestion
Like any development activity, the plan was to experiment with available tools. [doit](https://pydoit.org/) and [Typer](https://typer.tiangolo.com/) were the first ones we picked. The two components of our interests were a *task runner*, and a *command-line interface tool*. doit satisfied the requirements of a task runner along with added functionality, like maintaining a task dependency graph as a `DAG`. While `Typer` is quick to get started with building CLI applications.
```

Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:36]: I'm not sure if DAG is very familiar to people. Maybe link it to wiki?

**Comment by noatamir** [comment]
I left some suggestions and comments. I'm about half way through and need to stop for today.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 12 above)
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:36]: Sounds good. 
I found this article easy to understand - https://hazelcast.com/glossary/directed-acyclic-graph/ 
The wiki one seemed very detailed, https://en.wikipedia.org/wiki/Directed_acyclic_graph

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 10 above)
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:31]: It's not a SciPy lingo (maybe just me using the term), I shall update the wording. Thanks for pointing :)

By conditional blocks, I was trying to infer multiple `if-else` statements. Maybe replacing the term `conditional` with the `if-else` along with an example could be helpful?

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 9 above)
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:28]: Over time, the contribution guides start to span over more and more pages, `as` the over all effort to maintain the project grows exponentially

Thanks for the suggestion :)
Perhaps we could replace `as` with `and`, the result and outcome sound opposite here.

Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md]: both work!
`as` makes a causal links. the guide is growing because the the maintenance effort is grow. 
`and` just means both these things happen. no causality implied.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 10 above)
Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:31]: Oh yeah, I understand that the conditionals mean if-end statements. I just don't understand why that information is relevant to the story you are telling me. Is it particularly interesting? It just seems like a very specific detail to mention here that isn't connected to the rest of the paragraph. Are conditionals notoriously hard to maintain and I'm missing the point?

**Comment by noatamir** [comment]
continuing now 😊

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+#### ✍️ Planning and objective
+Like any development activity, the plan was to experiment with available tools. [doit](https://pydoit.org/) and [Typer](https://typer.tiangolo.com/) were the first ones we picked. The two components of our interests were a *task runner*, and a *command-line interface tool*. doit satisfied the requirements of a task runner along with added functionality, like maintaining a task dependency graph as a [DAG](https://hazelcast.com/glossary/directed-acyclic-graph/). While `Typer` is quick to get started with building CLI applications.
+
+As a starting point, I began experimenting with existing `dev.py` options, wrapped around multiple composite `doit` and `Typer` tasks. Both the [doit POC](https://github.com/sayantikabanik/scipy/blob/cli_poc/dodo.py) and [Typer POC](https://github.com/sayantikabanik/scipy/blob/cli_poc/cli.py) were built with similar principles.
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:38]: Before using POC for the first time. Write it as `Proof of concept (POC)` and then you can continue with the shorter version. (though some people prefer to be even more explicit and write `Proof of concept (from now on POC for short)` or something like this 🤷🏽‍♀️. I think it's ok to use the former for blogs.

Not sure I understand what's `multiple composite doit and Typer tasks`.  Feels like I really need a lot of cognitive load to get into what a CLI to figure it out and I 'd rather that you tell me in simpler words if I'm skimming this blog or I'm a beginner. 

Also you say that both POCs were built with similar principles, but you don't state principle**s** after it. So Am I supposed to infer it's this wrapping of something I didn't bother to understand in the line before? I feel like either I missed something, or something is missing and I can't tell 🤷🏽‍♀️

(I know I'm being a bit purposefully difficult but I'm trying to think like a casual blog reader and ppl that visit our blogs spend a couple of minutes reading, and even then it might be that some aren't as technical as you are).

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+
+As I progressed with the development of POCs using both tools, I experienced certain shortcomings. A better way to integrate an external library for exposing a CLI was the missing piece. **Eduardo Naufel Schettino** the author of doit developed an architecture combining the core elements of `doit` along with `click`. The journey in detail is captured under [issue-#133](https://github.com/rgommers/scipy/issues/133).
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:40]:
```suggestion
As I progressed with the development of POCs using both tools, I experienced certain shortcomings. A better way to integrate an external library for exposing a CLI was the missing piece. **Eduardo Naufel Schettino** the author of doit developed an architecture combining the core elements of `doit` along with `click`. The journey is captured  in detail under [issue-#133](https://github.com/rgommers/scipy/issues/133).
```

Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:40]: word order

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+
+#### 💁🏽‍♀️ More about the architecture and core components
+
+Combining these tools wasn’t a straightforward journey; after multiple iterations, we were able to achieve a stable state. Doit underwent updates to incorporate added functionality, helping the pieces come together live. Below are the core components for the `doit-click` based task definition.
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:44]:
```suggestion
Combining these tools wasn’t a straightforward journey; after multiple iterations, we were able to achieve a stable state. Doit underwent updates to incorporate added functionalities, helping the pieces come to life together. Below are the core components for the `doit-click` based task definition.
```

Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:44]: Were the doit updates at your request? or were you just impacted by their project development? or did you/scipy team contribute to their project in order for it better match your needs? 
Because any of those is a cool detail to add!

I'm assuming it was more than one functionality based on what you wrote so I amended it to plural.

You can also remove `to life` or `together`, (I think having both is a bit redundant), and it works just as well. I think I would prefer it without `to life` personally. It give me weird vibes personally 🧟  but it's totally fine 😉  people use it all the time.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 14 above)
Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:40]: Is this where you say you move from 2 POCs to only doit+click? I read on a bit and it seems like  the rest doesn't include Typer. I think it's not super clear. Maybe add something at the end of this like, `henceforth I continued with a doit+click POC, but it's not quite the end yet...`? feel free to rewrite.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+##### Class based Click command definition
+```python
+@cli.cls_cmd('test')
+class Test():
+    """Run tests"""
+    @classmethod
+    def run(cls):
+        print('Running tests...')
+```
+- Command may make use of a Click.Group context defining a `ctx` class attribute
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:60]: **A** command or **Any** command. I'm missing a determinator in the beginning.. 
It's ok in the next line because `command options` is plural (thought you can write `The command options`). 

Since both of these bullets describe changes you introduce in the next example/illustration/deepdive, consider adding a line saying, `as you can see in the next code chunk/snippet/following` or maybe tag your code illustrations? like fig1. fig2.  but code snippet 1, code snippet 2? I don't know - maybe it's overkill but right now it's a bit hard to follow for neurodiverse people (NDs) in my opinion. (which often means other people will also appreciate it if it was a bit more clear 😉).

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+            "commands": ["notes", "authors"],
+        },
+    ]
+}
+```
+
+#### 🎥 The developer CLI in action
+
+<script id="asciicast-U9l9VvklvEjXdEi1xYS4A7u5M" src="https://asciinema.org/a/U9l9VvklvEjXdEi1xYS4A7u5M.js" async></script>
+
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:120]: I <3 all of this section! so cool!

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+* `Documentation tasks`
+     - doc (Build documentation)
+     - refguide-check  (Run refguide check)
+
+* `Release tasks`
+    -  notes   (Release notes and log generation)
+    -  authors  (Task to generate list the authors who contributed within a given revision interval)
+* `Benchmarking tasks`
+    - bench & bench compare
+
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:143]: excellent! very clear. is the empty line in the bullet list a problem in MD rendering? I don't recall...

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+#### 👏🏽 Great Collaboration and teamwork
+
+From an idea to developing a successful POC, the journey was a great learning opportunity for me. Planning, coordination, teamwork and clear communication played a very important role. Huge thanks to **Ralf Gommers** and **Eduardo Naufel Schettino** for the amazing collaboration and support.
+
+As a newcomer to the *SciPy codebase*, it was a steep learning curve. I asked a ton of questions, at times drifted to the ocean searching for answers. Learning something completely new be overwhelming, different emotions brush past. Thanks to Ralf and **Pamphile Roy** for addressing my questions. The learnings and achievements will stay with me for a long-long time.
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:148]:
```suggestion
As a newcomer to the *SciPy codebase*, it was a steep learning curve. I asked a ton of questions, at times drifted to the ocean searching for answers. Learning something completely new can be overwhelming, different emotions brush past. Thanks to Ralf and **Pamphile Roy** for addressing my questions. The learnings and achievements will stay with me for a long-long time.
```

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+
+#### 😇 The possibilities next steps
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:150]:
```suggestion
#### 😇 The next steps
```

Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:150]: concise and clear.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+3. ⏩ Clear and concise examples to get started quickly
+4. ⏱ Reduction in the time spent navigating documentation
+
+With a great start comes possibilities. In the coming weeks, the CLI will become more mature and stable. After we receive wider usage and acceptance from the community, support for `dev.py` and `runtests.py` will be paused and `do.py` will be renamed to `dev.py`. The user documentation for the CLI components and usage will be made available for clear and concise understanding.
+To foster reusability, Eduardo has developed a package named `pydevtool`. The reusable elements will be incorporated into the SciPy developer CLI code. We will also be adding support for `act`, which will enable users to run GitHub CI jobs locally.
+
+#### 🙂 Parting thoughts
+
+Many thanks to the wonderful community for all the support and guidance.
+We are excited to collaborate with projects looking forward to adapting a similar developer command-line interface.
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:170]: Since I'm curious about this myself. Do you plan to document something about how might build a doit+click tool in general so that others might follow this template? It can go on Scientific-Python.org as maintainer documentation 😉 😉 😉 😉 😉 😉

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+> **Outcomes**
+
+1. 📜 Self-documentation and hierarchical help option
+2. 🧭 Easy to navigate and intuitive interface
+3. ⏩ Clear and concise examples to get started quickly
+4. ⏱ Reduction in the time spent navigating documentation
+
+With a great start comes possibilities. In the coming weeks, the CLI will become more mature and stable. After we receive wider usage and acceptance from the community, support for `dev.py` and `runtests.py` will be paused and `do.py` will be renamed to `dev.py`. The user documentation for the CLI components and usage will be made available for clear and concise understanding.
+To foster reusability, Eduardo has developed a package named `pydevtool`. The reusable elements will be incorporated into the SciPy developer CLI code. We will also be adding support for `act`, which will enable users to run GitHub CI jobs locally.
+
```
Comment (noatamir) [inline posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:166]: This all sounds great!! I would only encourage you to think if you can say something about how when do.py becomes dev.py it is officially no longer a POC. 👏 for the work. This is very cool. excellent blog and project!

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 10 above)
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:31]: Long chains of conditional statements degrade code readability. Makes it harder to find the right block to update/modify something. 

Each conditional block is designed to pick a specific user input and perform certain actions. 
`if user_input==--test ...perform x`
     `else:___`
 `if user_input==--bench ...perform y`
      `else:___`
  This spans to multiple other blocks.
  
  I can remove the conditional wording or try adding a similar example.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 18 above)
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:143]: Markdown I believe. I checked the preview it's rendered properly there

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 10 above)
Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:31]: Ah, the way you explain it now is much more clear to me than the original text. May I suggest you amend to incorporate your new phrasing:
`Both of these tools have long chains of conditional statements, **which** are notorious for reducing code readability drastically. **That** makes it difficult to maintain such code, because it is harder to find the right block to modify.`

Now that you explained I can also understand the original 2 lines easily tbh. But originally, I didn't get the connection between  the first line and the second line.  In the suggestion I made I try to make the connection a bit stronger. so the reader understands the conditionals are the cause of the problem. But feel free to rewrite in your style.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 15 above)
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md]: The collaboration with Eduardo provided visibility wrt to the design choices and any additional doit updates. As the project progressed the required requirements/additions were made.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 22 above)
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md:166]: Currently, I don't have an idea of when that might be feasible and do.py becomes the single CLI. Though I believe it would be a gradual shift after we receive wider acceptance from the community.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+#### 💁🏽‍♀️ More about the architecture and core components
+
+Combining these tools wasn’t a straightforward journey; after multiple iterations, we were able to achieve a stable state. Doit underwent updates to incorporate added functionality, helping the pieces come together live. Below are the core components for the `doit-click` based task definition.
+
+1. ✍️ Click based approach to input arguments/options/parameters
+2. ☑️ Base class to define doit task and/or click command
+3. 🏃 Method to execute a task `run()`
+4. 🌟 Additional utilities like `task dependency` and metadata definition using class attribute `TASK_META`
+
+##### Class based Click command definition
```
[the remark this answers is not in the document]
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md]: Made a tiny addition for more clarity 
Below are the core components for the doit-click based task definition along with an `illustration`.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 16 above)
Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md]: The tags render beautifully in my code editor, I checked in deploy preview it doesn't look well 😕
Screenshot 2022-05-09 at 11 10 30 AM

**Comment by sayantikabanik** [comment]
Thanks again @noatamir for the details feedback, I have updated the required sections of the blogpost.

File: posts/2022/05/the-evolution-of-the-scipy-developer-cli.md (hunk 16 above)
Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md]: @sayantikabanik based on https://getnikola.com/handbook.html#markdown I'm gathering that the markdown (MD) flavour for our blog is python MD which follows these syntax rules: https://daringfireball.net/projects/markdown/syntax. 

Alas, it looks like Python MD doesn't support MD tables, which I believe is what you did with your tags? But you should be able to insert a table html tag, if you want to keep it that way (search for the word `table` on that page for a simple example). Or you can change the style to an emphasis (double underscore maybe?) or subheading (just the next level down?) if you'd rather avoid html and stick to pure MD.

I hope my advice here is correct 🤞 since I haven't tried it out myself. I simply tried to find out what the right MD flavour for your tag to render 😉. Let me know how it goes...

Comment (sayantikabanik) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md]: Thanks, @noatamir 😊 I updated the HTML table tag. Though it wasn't standing out as I expected. Hence switching to `Blockquote`.

Comment (noatamir) [reply posts/2022/05/the-evolution-of-the-scipy-developer-cli.md]: thought maybe now it also works without saying `additional details` there? not sure...

**Comment by noatamir** [comment]
Let me know when you're ready to publish. Looks great to me!

**Comment by sayantikabanik** [comment]
> Let me know when you're ready to publish. Looks great to me!

Super excited, about my first blog with Quansight. Do I need to follow any additional steps for the deployment to though successfully?

**Comment by noatamir** [comment]
It's my first time squash-merging on the blog so we will both find out together 😄 
All the checks are green so I think it should be fine 👍 
I will do this a bit later when I have more time I need to prepare for a meeting.

**Comment by rgommers** [comment]
Yay, nice post 🎉!

@noatamir I'm not sure how, but something didn't go quite right here: you managed to do both a squash-merge (in commit 4177e2c1b2) and a regular merge of the branch after that (in commit 3ee0fa70). The normal way of doing this is to select "Squash and merge" for the green button, hit that, and then do nothing else. 

There's no other commits on top yet, so I think I'll fix up the history by resetting `main` to commit 4177e2c1b2, to get rid of the 29 extra commits.
